src/plugins/s390.c

```diff
@@ -398,3 +399,399 @@ gchar* bd_s390_zfcp_sanitize_lun_input (const gchar *lun, GError **error) {
 
     return fulllun;
 }
+
+/**
+ * bd_s390_zfcp_online:
+ * @devno: zfcp device number
+ * @wwpn: zfcp WWPN (World Wide Port Number)
+ * @lun: zfcp LUN (Logical Unit Number)
+ * @error: (out): place to store error (if any)
+ *
+ * Returns: whether a zfcp device was successfully switched online
+ */
+gboolean bd_s390_zfcp_online (const gchar *devno, const gchar *wwpn, const gchar *lun, GError **error) {
+    gboolean boolrc = FALSE;
+    gint rc = 0;
+    FILE *fd = NULL;
+    DIR *pdfd = NULL;
+    const gchar *zfcp_cio_free[4] = {"/usr/sbin/zfcp_cio_free", "-d", devno, NULL};
+    const gchar *chccwdev[4] = {"/usr/sbin/chccwdev", "-e", devno, NULL};
+
+    gchar *zfcpsysfs = "/sys/bus/ccw/drivers/zfcp";
+    gchar *online = g_strdup_printf ("%s/%s/online", zfcpsysfs, devno);
+    gchar *portdir = NULL;
+    gchar *unitadd = NULL;
+    gchar *failed = NULL;
+
+    /* part 01: make sure device is available/not on device ignore list */
+    fd = fopen (online, "r");
+    if (!fd) {
+        boolrc = bd_utils_exec_and_report_error (zfcp_cio_free, NULL, error);
+
+        if (!boolrc) {
+            fclose (fd);
+            g_free (online);
+            g_set_error (error, BD_S390_ERROR, BD_S390_ERROR_DEVICE,
+                         "Could not remove device %s from device ignore list.", devno);
+            return FALSE;
+        }
+        /* fd is NULL at this point, so try to open it again */
+        fd = fopen (online, "r");
+        g_free (online);
+
+        /* still no luck, fail */
+        if (!fd) {
+            g_set_error (error, BD_S390_ERROR, BD_S390_ERROR_DEVICE,
+                         "Could not open device %s even after removing it from" " the device ignore list.", devno);
+            return FALSE;
+        }
+    }
+
+    /* part 02: check to make sure/turn device online */
+    rc = fgetc (fd);
+    if (rc == EOF) {
+        /* there was some error checking the 'online' status */
+        g_set_error (error, BD_S390_ERROR, BD_S390_ERROR_IO,
+                     "Error checking if device %s is online", devno);
+        fclose (fd);
+        g_free (online);
```

Comment:
Took me a moment to understand what you meant, but I see now. So I'll certainly fix this, just leaving the `g_free()` call here where you've pointed it out, but I'll only call `g_free()` after the `fopen()` retry if there's an error, which should be perfectly fine.